rfc/rfc-0000-k8s-events.md

```diff
@@ -0,0 +1,162 @@
+# Meta
+
+[meta]: #meta
+
+- Name: Introduce Kubernetes Events
+- Start Date: 2022-03-28
+- Author(s): [Rasheed Abdul-Aziz](https://github.com/squeedee)
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- Supersedes: N/A
+
+# Summary
+
+[summary]: #summary
+
+Introduce the use
+of [K8s Events](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#events)
+to provide user's awareness of state-changing event's that are meaningful for debugging and situational awareness.
+
+# Motivation
+
+[motivation]: #motivation
+
+Developers and Operators need to debug issues where workloads stall, do not make it to deployment, or where other level
+changes do not propagate as they should.
+
+Cartographer concern's itself with creating **Stamped Objects** from **Templates** and passing **Artifact**
+references between them. By design, the passing of references (and desired state definitions in the **owner** spec)
+is _by value_. A user can only see if a state change is transferred from source to sink, not when it happened, or how
+often, or even what the specific cause was for a resource to be updated.
+
+This approach lends itself well to the declarative, eventual consistency model of Kubernetes, where "I want this, I
+don't care how you get me there" reigns. However, there are times when a user needs to know more detail about the
+machinations of the Cartographer controller, especially in debugging failures. For these situations, Kubernetes
+provides [events](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#events)
+to expose the user to these temporal triggers/actions from inside controller processes.
+
+**Stamping** (creation of the resource definition from template and inputs) and **Submitting** (applying the resource
+definition to the cluster) are two very common events, with associated causes, that currently need further exposure to
+users.
+
+# What it is
+
+[what-it-is]: #what-it-is
+
+## Developer View
+
+Cartographer presents useful information to developers in the **owner** (workload/deliverable). This information is
+derived from stamped objects and contributing resources such as **templates** and **blueprints**:
+
+This RFC addresses the absence of awareness that "something is happening", especially after a developer has committed
+new code. With event's we can provide information that show's activity and causality not present in the current snapshot
+of state.
+
+We add events to the **Owners** (Workload/Deliverable) to enable debug and reasoning about Cartographer's behavior.
+
+## Operator View
+
+The situation is worse for users managing the **blueprints** (Devops/Ops). Some metrics can be recorded in **
+blueprints**
+and this might be a valid topic for another RFC, however events also present a useful way for authors to avoid trawling
```

Comment:
indeed, one could certainly make the argument that an ops team that makes use of any solution that converts those in-cluster events to data points to derive insights in some form could make use of these owner-based events for monitoring in some capacity

Reply:
It's pretty commonplace for admins to use monitoring tools with nice temporal graphs showing histograms that highlight usage, network partitioning events, heavy load events, etc.